codespan-reporting/src/term/utils.rs

```diff
@@ -0,0 +1,77 @@
+use std::ops::Range;
+
+pub fn closest_char_boundary(s: &str, byte_index: usize) -> usize {
+    let length = s.len();
+    for index in byte_index..=length {
+        if s.is_char_boundary(index) {
+            return index;
+        }
+    }
+    length
+}
+
+pub fn closest_char_boundary_rev(s: &str, byte_index: usize) -> usize {
```

Comment:
Why do we need a different algorithm for the end bound?

Reply:
Essentially searches for the closest boundary from the current index moving towards the left, (closest_char_boundary_rev), and the other goes from left to right for the next available boundary, both return the current character if it's a boundary, With the same algorthm for start and end I believe it would always return an empty string.

Reply:
Wouldn't that be fine, given that we'll still render a single caret for that?

Reply:
Hmm, I think this is fine though, was just asking! Can we put these functions and tests in the `renderer` module for now?

Reply:
It might be, I think isn't it also used for displaying the source code text, or is that a separate line range?

Reply:
sure I'll move.

Reply:
So, moved into place, while I was at it I added some doc strings.